resource/readers/resource_reader_rv1exec.cpp

```diff
@@ -222,6 +222,239 @@ int resource_reader_rv1exec_t::add_cluster_vertex (resource_graph_t &g,
+    auto vtx_iter = m.by_path.find (path);
+    if (vtx_iter != m.by_path.end ()) {
+        for (vtx_t v : vtx_iter->second) {
+            if (g[v].rank == rank
+                    && g[v].id == id
+                    && g[v].size == size
+                    && g[v].type == type) {
+                vtx = v;
+                break;
+            }
+        }
+    }
+
+    return vtx;
+}
+
+int resource_reader_rv1exec_t::update_vertex (resource_graph_t &g,
+                                              vtx_t vtx,
+                                              updater_data &update_data)
+{
+
+    int rc = -1;
+    int64_t span = -1;
+    int64_t avail = -1;
+    planner_t *plans = NULL;
+
+    // Check and update plan
+    if ( (plans = g[vtx].schedule.plans) == NULL) {
+        errno = EINVAL;
+        m_err_msg += __FUNCTION__;
+        m_err_msg += ": plan for " + g[vtx].name + " is null.\n";
+        goto error;
+    }
+    if ( (avail = planner_avail_resources_during (plans,
+                                                update_data.at,
+                                                update_data.duration)) == -1) {
+        m_err_msg += __FUNCTION__;
+        m_err_msg += ": planner_avail_resource_during return -1 for ";
+        m_err_msg += g[vtx].name + ".\n";
+        goto error;
+    }
+    if (avail < g[vtx].size) {
+        // if g[v] has already been allocated/reserved, this is an error
+        m_err_msg += __FUNCTION__;
+        m_err_msg += ": " + g[vtx].name + " is unavailable.\n";
+        goto error;
+    }
+    // Update the vertex plan here (not in traverser code).
+    // Traverser update () will handle aggregate filters and
+    // exclusivity checking filter.
+    // Can't update the rank-level (node) vertex yet- we
+    // don't know if all its children are allocated.
+    // Note this is a hard-coded option. Support for more flexible
+    // types may require extending rv1exec.
+    if (g[vtx].type != "node") {
+        if ( (span = planner_add_span (plans, 
+                                       update_data.at,
+                                       update_data.duration,
+                            static_cast<const uint64_t> (g[vtx].size))) == -1) {
+            m_err_msg += __FUNCTION__;
+            m_err_msg += ": can't add span into " + g[vtx].name + ".\n";
+            goto error;
+        }
+        
+        if (update_data.reserved)
+            g[vtx].schedule.reservations[update_data.jobid] = span;
+        else
+            g[vtx].schedule.allocations[update_data.jobid] = span;
+
+        update_data.updated_vertices[g[vtx].rank].push_back (vtx);
+    }
+ 
+    rc = 0;
+
+error:
+    return rc;
+}
+
+int resource_reader_rv1exec_t::undo_vertices (resource_graph_t &g,
+                                              updater_data &update_data)
+{
+
+    int rc = -1;
+    int64_t span = -1;
+    planner_t *plans = NULL;
+
+    for (auto rank : update_data.updated_vertices) {
```

Comment:
This should be a reference, we don't want to copy the vector of vertices each iteration.

```suggestion
    for (auto &rank : update_data.updated_vertices) {
```

Reply:
Ideally also name the parts:

```suggestion
    for (auto &[rank,vertices] : update_data.updated_vertices) {
```

Reply:
I like the the structured bindings approach you recommended but got a compile warning:
```bash
structured bindings only available with ‘-std=c++17’ or ‘-std=gnu++17’
```
since cmake is setting `-std=c++14` in bookworm for some reason. I'm going to push this change and see what happens with the other CI images.

Reply:
Are you rebased on the cxx20 update? That warning should have gone away, if not I missed something over there.

Reply:
Yes, I am. This appears to be happening locally. The `bookworm` in GitHub actions set the standard to 20 (as did all the other CI builds that I checked).